app/services/ab_test_bucket.rb

```diff
@@ -0,0 +1,65 @@
+class AbTestBucket
+  include ActiveModel::Model
+
+  attr_reader :buckets
+
+  validate :within_100_percent
+
+  def initialize(buckets: { default: 100 })
+    @buckets = buckets
+    raise 'invalid bucket data structure' unless valid_bucket_data_structure?
+    ensure_numeric_percentages
+    raise 'bucket percentages exceed 100' unless valid?
+  end
+
+  def bucket(discriminator = nil)
+    return :default if discriminator.blank?
+
+    user_value = percent(discriminator)
+
+    min = 0
+    buckets.keys.each do |key|
+      max = min + buckets[key]
+      return key if user_value > min && user_value <= max
+      min = max
+    end
+
+    :default
+  end
+
+  private
+
+  def percent(discriminator)
+    max_sha = (16 ** 64) - 1
+    Digest::SHA256.hexdigest(discriminator).to_i(16).to_f / max_sha * 100
```

Comment:
oh one more idea: we should inject something else into here (like a slug name for the test) for additinoal randomness. That way, if a user's plain UUID hashes to like 1%, they're not enabled in every single test we do

```suggestion
    max_sha = (16 ** 64) - 1
    Digest::SHA256.hexdigest("#{discriminator}:#{experiment_name}").to_i(16).to_f / max_sha * 100
```

Reply:
Love it - done in 9b3a258e7